src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.ProjectRestoreInfoSource.cs

```diff
@@ -144,13 +144,16 @@ private void SaveNominatedConfiguredVersions(IReadOnlyCollection<PackageRestoreC
                 {
                     _savedNominatedConfiguredVersion[configuredInput.ProjectConfiguration] = configuredInput.ConfiguredProjectVersion;
                 }
+            }
+        }
 
-                if (_whenNominatedTask is not null)
+        private void CompleteTaskWhenNominated()
+        {
+            if (_whenNominatedTask is not null)
```

Comment:
This used to be protected by a lock, now it is not. This has a race, by the time you've run TrySetResult this field could be null.

Reply:
yeah, I don't understand this is the fix. Just like David said, this must be in a lock, and also inside the same lock region when _saveNorminatedConfiguredVersion is updated.  There is no logic reason why this is a fix to the problem. Do we know the exact reason why nominated task is completed, because we return an expired task, or we complete it very quickly.

One possibility of this is that NuGet actually ask other projects when first project nominates, and then wait on the second project, which will nominate in a short time. Basically, it is chasing nominations in the system. Is that a possible explanation of this.  If that is the case, it might be expected?

Reply:
NuGet iterates over the known project checking HasPendingNomination. 
If any project returns true, then we call WhenNominated. 
NuGet will wait on at most one project at any given time. 

https://github.com/dotnet/project-system/issues/7453 has some logs of all nuget calls to the project system.